File: solutions/0383. Ransom Note/0383.cpp
// Time complexity: O(n+m)
// Space complexity: O(1)
class Solution {
 public:
  bool canConstruct(string_view ransomNote, string_view magazine) {
    vector<int> count(26);

    for (const char c : magazine)
      ++count[c - 'a'];

    for (const char c : ransomNote) {
      if (count[c - 'a'] == 0)
        return false;
      --count[c - 'a'];
    }

    return true;
  }
};

File: solutions/0383. Ransom Note/0383.py
# Time complexity: O(n+m)
# Space complexity: O(1)
class Solution:
  def canConstruct(self, ransomNote: str, magazine: str) -> bool:
    count1 = collections.Counter(ransomNote)
    count2 = collections.Counter(magazine)

    return all(count1[c] <= count2[c] for c in string.ascii_lowercase)